### ISSUE TITLE
[setters / firebase] events & hooks

### ISSUE THREAD

**Issue by mesqueeb** [vuex-cloud/rfcs#8]

**Comment by mesqueeb** [comment]
@louisameline
> Also, we'll talk about the signature, passing an object is more generic than a finite number of arguments. We might have more things to add in the future. Also, I think it should be like event in the DOM: if you **don't do** anything (like calling `preventDefault`), everything continues normally. We could have several functions to delay or abort for example, maybe others.

> Not sure if we actually need to expose these events. The `mutators` I think of will do the job of conversion. I'd have to think more to see if these are really useful

> A trick to avoid "beforeUpdate" and "afterUpdate" is to name them "update" (verb, something wants to do something) and "updated" (the action has been carried out). Works most of the time, not with irregular verbs like `set` though

**Comment by mesqueeb** [comment]
So I've been thinking of the internal workings.

These are the points we need hooks / events to be triggered:

<img width="1242" alt="Screenshot 2019-11-30 at 13 44 51" src="https://user-images.githubusercontent.com/3253920/69895646-aa743580-1377-11ea-8a51-e0c7ce32b138.png">

In v-e-f the local insert action will update the Vuex store via mutation and then execute another action returns the promise from the Firestore action.
In other words:
- It was easy for me to execute both the the local update then dispatch the sync action
- It was easy for me to just check for a hook function in the config, and execute this one if it exists

If we create an "event dispatcher" it's a radically different way from how v-e-f did it, and I'll have to think about how this is possible.

Here I divided the same scheme but added what is who's responsibility:

<img width="1308" alt="Screenshot 2019-11-30 at 13 40 39" src="https://user-images.githubusercontent.com/3253920/69895771-376bbe80-1379-11ea-8053-1fb9473ef097.png">

#### Things to keep in mind:

- at any point the payloads should be mutatable and then passed onto the next step (to allow adding default values etc.)
- at any point the chain should be abortable

**Comment by mesqueeb** [comment]
My post above is easily achieved with my proposed syntax [here](https://github.com/vuex-cloud/rfcs/issues/11#issuecomment-569467055).
All hooks are triggered by the core library.

**Comment by louisameline** [comment]
Note: hooks are not a good solution because:

1. They have to sit in the definition of your module, away from the rest of the data handling logic which is either in the app methods or in the store actions.
Personally, I have actions, state and getters in 3 different dedicated files (not in the module definition file), so they're not even in the same file as my hooks, which are left alone there.

2. There can be only one function per hook, which has to mix several concerns. If you have several different things to do at the same time of the processing cycle, they must be written together in the same function.

3. They are hardly modular. The same function will be run for all requests, so if some actions require something and the others do not, you'll have to write a lot of conditions in your hook callback.

```javascript
// in the module definition
serverChange: {
  addedHook: function (updateStore, doc, id, store) {

    const modify = () => { 'some data handling logic that I wish was somewhere else' }
  
    // you'll see later why I write this here. Imagine it's something that must always run
    doAlways()

    // this on the other hand is conditional
    if (iAmDoingThisKindOfThing === true) {
      doThis()
      updateStore(modify(doc))
    }
    // actually I'd use a switch or whatever, but just for clarity
    else if(doingSomethingElse) {
      doThat()
      // this PR needs acceptance by the way ;)
      updateStore(false)
    }
    ...
  }
}
```
becomes
```javascript
// neatly sitting among my other methods:
modify () { 'some data handling logic' }

// within the rest of my app logic, when I'm DoingThisKindOfThing 
this.$store.dispatch('module/fetch', {
  // this syntax can be improved, I'll discuss this down below
  on: event => {
    doAlways()
    if (event.type === 'added') {
      doThis()
      // this is how you can modify the payload when using events
      // (the syntax might change but you get the idea)
      event.updateStore(modify(event.doc))
    }
  }
})

// in some other place where I'm doingSomethingElse
this.$store.dispatch('module/fetch', {
  on: event => {
    doAlways()
    if (event.type === 'added') {
      doThat()
      // and this is how you stop the execution
      event.updateStore(false)
    }
  }
})
```
And voilà, things sit where they belong. More in the spirit of Vue3 or whatever.

However the  `on: event => {}` syntax basically forces us to write an `if` or a `switch` within the function.
This could be better:
```javascript
this.$store.dispatch('module/fetch', {
  listen: emitter => {
    emitter.on('added', event => {
      doAlways()
      doThat()
      event.updateStore(false)
    }
  }
})
```
It's easier this way to see what happens, and it's also easier to split your code in more subfunctions, like 
```javascript
this.$store.dispatch('module/fetch', {
  listen: emitter => {
    doAlways()
    // easily bind on component methods
    emitter
      .on('added', this.added)
      .on('fetched', this.fetched)
  }
})
```
Also, I like that a method called `on` get an event type as first parameter. It's how it works in the rest I describe below.

So now this is nice, and lets you define what you need for each action you call. But you could say "Hey what if I don't want to bind my listeners again everytime I call an action in my code? My requests do the same thing most of the time. I'd like to write some of my listeners in a single centralized place". Yes sure, that was an advantage of the hooks, which were centralized as well.

That's when the module event emitter comes into play, I mentionned it earlier in one of our conversations. Here's the trick: we can listen to the events from within a particular action call like above, or at module level. That's why I've written the `doAlways()` function up there. So now we can write the above block like this:

```javascript
// listen at module level, in a centralized manner, for generic stuff
this.vuexSyncModule.emitter.on('added', event => {
  doAlways()
})

// listen at action call level, for more specific things that happen just here
this.$store.dispatch('module/fetch', {
  listen: emitter => {
    // we don't need to call `doAlways()` here now, it will be taken care of
    // when the event reaches our other listener at module level. We just write
    // our call-specific stuff:
    emitter.on('added', this.added)
  }
})
```
The only difference with the previous block is that `doAlways()` will be called after `this.added` because the listeners at action call level should be called before the listeners at module level, to respect the natural "bottom-to-top" way events usually go (bubbling vs capturing, which might be considered if someone had a valid case).

So, with this we no longer need hooks.

**Comment by mesqueeb** [comment]
Yes I agree let's only use events in the manner you described above.

**Comment by mesqueeb** [comment]
@louisameline I'm tinkering a bit with the syntax we could use for core, unrelated of vuex or firestore.

We need some param called `next` or something that devs can decide not to call when they want to abort.

I believe that these are the events we need to capture:
```js
// local events
const beforeLocalUpdate = localOrRemote === 'local' && status === 'before'
const onLocalUpdateSuccess = localOrRemote === 'local' && status === 'success'
const onLocalUpdateError = localOrRemote === 'local' && status === 'error'

// remote events
const beforeRemoteUpdate = localOrRemote === 'remote' && status === 'before'
const onRemoteUpdateSuccess = localOrRemote === 'remote' && status === 'success'
const onRemoteUpdateError = localOrRemote === 'remote' && status === 'error'
```
The order of execution of these would be dependant on the sync paradigm: local  → remote or remote → local.

You mentioned:
> listen at module level, in a centralized manner, for generic stuff

I totally agree. let's say, there is a centralised event listeners the dev can set up. Besides this, when it comes to the action-per-action event listeners, there are so many ways we can write this events syntax, and I don't see which one is best or for what reason...

### 1. with emitter.on
```js
const { id } = usersModule.insert(
  { name: 'luca' }, // payload
  {
    listen: emitter => {
      emitter.on('beforeLocalUpdate', ({payload, next}) => {})
      emitter.on('onRemoteUpdateSuccess', ({payload, next}) => {})
    },
  }
)
```
### 2. with if statements & single event name
```js
const { id } = usersModule.insert(
  { name: 'luca' }, // payload
  {
    on: ({event, payload, next}) => {
      if (event === 'beforeLocalUpdate') {}
      if (event === 'onRemoteUpdateSuccess') {}
    },
  }
)
```
### 3. with if statements & combined event name
```js
const { id } = usersModule.insert(
  { name: 'luca' }, // payload
  {
    on: ({target, status, payload, next}) => {
      if (target === 'local' && status === 'before') {}
      if (target === 'remote' && status === 'success') {}
    },
  }
)
```

I think you'd prefer 1, i'm leaning towards 2, it seems simpler.

When written like this, I don't see any particular advantage for either 1 or 2, they both can do the same things, so it's just personal preference. Please let me know your thoughts and if there's a reason 1. would be better for a specific reason. 😄

**Comment by louisameline** [comment]
Hi Lucas! If we had wanted `on` to be a function at action call level, I would have wanted it to have the same syntax as our `SyncModule.on('eventType', fn)` module level feature. But an `on` object is fine here you're right, sold! ^^

Not related, but I need to take some time this week to talk to you about a potential feature to make structural changes to the store easier to carry out.
